runtime/grid_reduction.cu

```diff
@@ -667,4 +667,243 @@ __device__ void serialReductionStep(
+    int vec_size,
+    typename T,
+    typename Func>
+__device__ void iterGroupedGridReduce(
+    T* out,
+    const T* inp_val,
+    Func reduction_op,
+    volatile T* work_buf,
+    int64_t* sync_flags,
+    T* shared_buf,
+    bool read_pred,
+    bool write_pred,
+    T init_val,
+    const nvfuser_index_t entrance_ind,
+    const nvfuser_index_t n_entrances) {
+  // inp or block reduction results
+  T block_reduction_val[vec_size];
+
+  // Do block reduction when required
+  if (X_THREAD || Y_THREAD || Z_THREAD) {
+#pragma unroll
+    for (int i = 0; i < vec_size; i++) {
+      block_reduction_val[i] = init_val;
+    }
+    blockIterGroupedReduce<X_THREAD, Y_THREAD, Z_THREAD, Aligned, vec_size>(
+        block_reduction_val,
+        inp_val,
+        reduction_op,
+        shared_buf,
+        read_pred,
+        true,
+        init_val);
+  } else if (read_pred) {
+#pragma unroll
+    for (int i = 0; i < vec_size; i++) {
+      block_reduction_val[i] = inp_val[i];
+    }
+  }
+
+  // Number of values to reduce in the reduction segment
+  const auto grid_reduction_segment_size =
+      index_utils::maskedSize<X_BLOCK, Y_BLOCK, Z_BLOCK>(gridDim);
+
+  // Index of the reduction we're performing out of the
+  // grid_reduction_segment_size
+  const auto idx_in_grid_segment =
+      index_utils::maskedOffset<!X_BLOCK, !Y_BLOCK, !Z_BLOCK>(
+          blockIdx, gridDim);
+
+  // Number of threads we can use in final reduction, Seems to assume all
+  // threads in the block participate
+  const auto block_reduction_segment_size =
+      index_utils::maskedSize<!X_THREAD, !Y_THREAD, !Z_THREAD>(blockDim);
+
+  // Number of reductions in the grid
+  const nvfuser_index_t grid_segment_size = PERSISTENT_REDUCTION
+      ? 1
+      : index_utils::maskedSize<!X_BLOCK, !Y_BLOCK, !Z_BLOCK>(gridDim);
+
+  // advance to the offset for this segment
+  // index of reduction * size of the reduction * size of threads
+  if ((!X_THREAD || threadIdx.x == 0) && (!Y_THREAD || threadIdx.y == 0) &&
+      (!Z_THREAD || threadIdx.z == 0)) {
+    auto block_offset =
+        index_utils::maskedOffset<X_BLOCK, Y_BLOCK, Z_BLOCK>(blockIdx, gridDim);
+    auto thread_offset =
+        index_utils::maskedOffset<!X_THREAD, !Y_THREAD, !Z_THREAD>(
+            threadIdx, blockDim);
+
+    // vectorized write to work_buf [global memory]
+    auto work_buf_offset =
+        block_offset * (block_reduction_segment_size * grid_segment_size) +
+        block_reduction_segment_size * idx_in_grid_segment + thread_offset;
+    work_buf_offset *= vec_size;
+
+    constexpr int total_bytes = vec_size * sizeof(T);
+    constexpr int n_loads = total_bytes <= 16 ? 1 : total_bytes / 16;
+    constexpr int n_elements = total_bytes <= 16 ? vec_size : 16 / sizeof(T);
+#pragma unroll
+    for (int i = 0; i < n_loads; i++) {
+      loadLocalToGlobal<T, n_elements, true>(
+          &work_buf[work_buf_offset + i * n_elements],
+          &block_reduction_val[i * n_elements]);
+    }
+  }
```

Comment:
I notice that this kernel is slightly different from gridReduction:

We are not updating work_buf here. `thread_offset` computed in this function is also different from the one in the last block. For my own curiosity, wondering what's the reason behind the change?

Reply:
I changed how the block reduction result is written to gmem. It ensures contiguous vectorized access.
`work_buf` is the base pointer, `work_buf_offset` is different for each thread to ensure we are writing to the correct gmem. I added additional comments.

Reply:
For `thread_offset`, added comments:
```  
// Its calculation is same to that in [iterGroupedGridReduce]. Becuase when
  // [iterGroupedGridReduceLastBlock] is called from [iterGroupedGridReduce],
  // X_THREAD, Y_THREAD, Z_THREAD are flipped.
``
We can also make it work without doing the flip in the function call but I didn't have a good reason do that change.

Reply:
Ha I didn't realized that we flipped the flags. No worries.

I think the old logic is that, if we increment `work_buf` in the original function with `thread_offset`, then we don't have to compute it again in `last block`. But that's probably nitpicking too much in the code. We can revisit this when we merge the two functions.